ddtrace/internal/stats.py

```diff
@@ -0,0 +1,150 @@
+        """Internal helper to get the current value of a counter since last check"""
+        key = self._key(name, metric_type, tags)
+
+        # Get the current value and last value we saw
+        val = self._counters[key].value(no_lock=True)
+        last_value = self._last_values[key]
+
+        # For increments, only grab the difference between
+        #   last time we fetched and now
+        # For gauges, we always want the current total value
+        if metric_type is self.METRIC_TYPE_INCREMENT:
+            # Compute the change in value since last check
+            val = val - last_value
+
+        # Store the current value for next time we fetch
+        self._last_values[key] = val
+        return val
+
+    def reset_values(self):
+        """
+        Return and reset the current value of all counters
+
+        ::
+
+            from ddtrace.internal.stats import get_stats
+
+            # Get global stats instance
+            stats = stats.get_stats()
+
+            # Increment counters
+            stats.span_started()
+            stats.span_finished()
+
+            # Fetch all current metrics, resetting their internal values back to 0
+            for metric_name, metric_type, value, tags in stats.reset_values():
+                pass
+
+        :returns: List of ``(metric_name, metric_type, value, tags)`` for each stat monitored
+        :rtype: :obj:`list`
+        """
+        with self._read_lock:
+            # Collect and reset all current counters
+            values = []
+            for name, metric_type, tags in self._counters.keys():
+                val = self._get_value(name, metric_type, tags)
+                # Convert tags to a list so we can add to default tags set on the dogstatsd client
+                if tags is not None:
+                    tags = list(tags)
+
+                values.append((name, metric_type, val, tags))
+
+            return values
+
+    def report(self, dogstatsd_client):
+        """
+        Report all existing metrics to the provided dogstatsd client
+
+        :param dogstatsd_client: A DogStatsd client to send metrics to
+        :type dogstatsd_client: :class:`ddtrace.vendor.dogstatsd.DogStatsd`
+        """
+        for metric, metric_type, value, tags in self.reset_values():
+            getattr(dogstatsd_client, metric_type)(metric, value, tags=tags)
+
+            # Manually increment a "sum" counter for histograms
+            # DEV: `<metric>.sum` histogram metric is disabled by default in the agent
+            if metric_type is self.METRIC_TYPE_HISTOGRAM:
+                dogstatsd_client.increment("{}.total".format(metric), value, tags=tags)
+
+
+# Default global `Stats` instance
+# Get this from `ddtrace.internal.stats.get_stats`
+_stats = Stats()
+
+
+def get_stats():
+    """
+    Fetch the current global :class:`Stats` instance
+
+    :returns: Global :class:`Stats` instance
+    :rtype: :class:`Stats`
+    """
+    global _stats
+    return _stats
```

Comment:
I think I know the answer but I still need to ask: what about per-`Tracer` stats?

Reply:
I think there are some stats we could track which are not related to a tracer, or do not require a tracer. (lets say we wanted to keep track of the number of tracers that are created, etc).

Also, the writer doesn't have access to the tracer, so we'd need to know about all the `Stats` objects globally to report stats.